[felixchanyy] iP

felixchanyy opened the pull request:
# Denify

Zenify brings peace to your mind by relieving you of the burden of remembering tasks. It's,

- Text-based
- Easy to learn
- FAST SUPER FAST to use

All you need to do is:

1. Download it from [here](https://github.com/felixchanyy/ip/).
2. Start using it.
3. Let it manage your tasks for you 😉

And it is **FREE**!

## Features:

- [x] Managing todo
- [x] Managing deadlines
- [x] Managing events

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        new Duke("./data/duke.txt").run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

felixchanyy committed:
Added rename chatbot, greet and exit implementation

felixchanyy committed:
Add echos command implementation

felixchanyy committed:
Add new task to list and display back

felixchanyy committed:
Add the ability to mark tasks as done or not done (both ways)

felixchanyy committed:
Add ToDos, Events, Deadlines implementation

felixchanyy committed:
Add semi-automate the testing of Duke

felixchanyy committed:
Add handle errors

felixchanyy committed:
Add support for deleting tasks from the list

felixchanyy committed:
Add Java enums

felixchanyy committed:
Add the tasks in the hard disk automatically

Save the tasks in the hard disk automatically whenever the task list
changes. Load the data from the hard disk when the chatbot starts up.

felixchanyy committed:
Merge branch 'branch-Level-7'

felixchanyy committed:
Teach the chatbot how to understand dates and times

Store deadline and event dates as java.time.LocalDateTime in
my task objects.

Accept dates in a format such as yyyy-mm-dd HH:mm format
(e.g., 2019-10-15 17:00) and print in a different format
such as MMM dd yyyy HH:mm e.g., (Oct 15 2019 17:00).

felixchanyy committed:
Merge branch 'branch-Level-8'

felixchanyy committed:
Make the code more OOP and fix date time format

Make the code more OOP by the creation of the following classes.

Ui: deals with interactions with the user.

Storage: deals with loading tasks from the file and saving tasks
in the file.

Parser: deals with making sense of the user command.

TaskList: contains the task list e.g., it has operations to
add/delete tasks in the list.

felixchanyy committed:
Merge branch 'branch-A-MoreOOP'

felixchanyy committed:
Divide classes into packages

The codebase needs better organization, and introducing packages
will improve maintainability and structure.

Organize classes into packages according to their functionality and
purpose.

Using packages aligns with best practices for Java projects,
enhancing code readability and maintainability.

This change ensures a more structured and scalable project
architecture.

felixchanyy committed:
Merge branch 'branch-A-Packages'

felixchanyy committed:
Merge branch 'add-gradle-support'

felixchanyy committed:
Add Gradle

This project currently lacks a standardized build system, making it
difficult to manage dependencies and build processes consistently.

Introduce Gradle as the build tool for the project to provide a
well-organized and maintainable structure. This change facilitates
dependency management, build automation, and enhances the overall
project development experience.

Gradle is chosen for its flexibility, plugin ecosystem, and ease of use.
It ensures consistent builds across different environments and
simplifies the management of project dependencies.

This commit sets up the basic Gradle configuration.

felixchanyy committed:
Merge branch 'branch-A-Gradle'

felixchanyy committed:
Add JUnit tests for Parser and TaskList classes

The current situation lacks sufficient test coverage for the Parser
and TaskList classes, particularly for their non-trivial methods.

For the Parser class, JUnit tests have been introduced to cover
various scenarios and ensure the correctness of parsing methods.
This approach enhances the reliability of the Parser class,
making it more robust and resilient to different input formats.

For the TaskList class, comprehensive JUnit tests have been added
to cover key functionalities such as adding, deleting, marking, and
unmarking tasks. This testing approach ensures that the TaskList class
behaves correctly under different scenarios, catching potential bugs
and enhancing the stability of the application.

No other relevant information at this time.

felixchanyy committed:
Merge branch 'branch-A-JUnit'

felixchanyy committed:
Configure executable JAR file.

To make the application distribution more convenient for users.

Configure the project to generate an executable JAR file.

Packaging the application as an executable JAR simplifies
distribution and allows users to run the application easily
with a simple command.

Users can now run the application by executing
`java -jar Duke.jar`

felixchanyy committed:
Merge branch 'branch-A-Jar'

felixchanyy committed:
Add JavaDocs

JavaDoc comments are essential for providing
clear documentation about classes, methods, and other code elements.
They serve as a valuable resource for developers, enhancing code
understanding, and facilitating maintenance.

Added JavaDoc comments to all classes
and methods, ensuring that each comment follows the established
coding standards. The comments include relevant information about
the purpose, parameters, return values, and exceptions of the methods.

Following the established JavaDoc
conventions ensures consistency and readability. The comments
are concise, providing essential information without unnecessary
details. The goal is to improve code documentation and make it
easier for developers to understand and work with the codebase.

Developers are encouraged to maintain
the JavaDoc comments whenever they make changes to the code to
keep the documentation up-to-date.

felixchanyy committed:
Merge branch 'branch-A-JavaDoc'

felixchanyy committed:
Merge branch 'branch-A-JavaDoc' into branch-A-CodingStandard

felixchanyy committed:
Adhere to Coding Standard

Improve method names in ParserTest and TaskListTest
for clarity and consistency.

Refactor to align with coding standard.

felixchanyy committed:
Merge branch 'branch-A-CodingStandard'

felixchanyy committed:
Merge branch 'branch-A-CodingStandard' into branch-Level-9

felixchanyy committed:
Give users a way to find a task by searching for a keyword

Improve task search functionality

Users should have the ability to find a task by searching for a keyword.
This commit introduces enhancements to the search feature.

- Implement a method in the TaskList class to find tasks containing
a specified keyword.
- Update the Parser class to handle the "find" command and parse
the keyword.
- Enhance the UI to display matching tasks when the "find"
command is used.

This approach provides a more user-friendly experience,
allowing users to quickly locate tasks based on specific keywords.

felixchanyy committed:
Merge branch 'branch-Level-9'

felixchanyy committed:
Add addTask when user enter todo, event or deadline

Forgot to add the line where task will be added to tasklist

limrik commented on src/main/java/duke/core/Duke.java:
```diff
@@ -0,0 +1,128 @@
+                    ui.formatReply(ui.exit());
+                    break;
+                } else if (msg.toUpperCase().startsWith(Command.LIST.name()) && parser.parseList()) {
+                    ui.formatReply(ui.showAllTasks(tasks.getTasks()));
+                } else if (msg.toUpperCase().startsWith(Command.FIND.name())) {
+                    String description = parser.parseFind();
+                    ui.formatReply(ui.showFoundTasks(tasks.findTasks(description)));
+                } else if (msg.toUpperCase().startsWith(Command.MARK.name())) {
+                    int taskIndex = parser.parseMark();
+                    Task t = tasks.markTask(taskIndex);
+                    ui.formatReply(ui.showMarkTaskMessage(t));
+                    tasks.saveToStorage(storage);
+                } else if (msg.toUpperCase().startsWith(Command.UNMARK.name())) {
+                    int taskIndex = parser.parseUnmark();
+                    Task t = tasks.unmarkTask(taskIndex);
+                    ui.formatReply(ui.showUnmarkTaskMessage(t));
+                    tasks.saveToStorage(storage);
+                } else if (msg.toUpperCase().startsWith(Command.DELETE.name())) {
+                    int taskIndex = parser.parseDelete();
+                    Task t = tasks.deleteTask(taskIndex);
+                    ui.formatReply(ui.showDeleteTaskMessage(t, tasks.getTasks().size()));
+                    tasks.saveToStorage(storage);
+                } else if (msg.toUpperCase().startsWith(Duke.Command.TODO.name())) {
+                    Task t = parser.parseTodo();
+                    tasks.addTask(t);
+                    ui.formatReply(ui.showAddTaskMessage(t, tasks.getTasks().size()));
+                    tasks.saveToStorage(storage);
+                } else if (msg.toUpperCase().startsWith(Command.DEADLINE.name())) {
+                    Task t = parser.parseDeadline();
+                    tasks.addTask(t);
+                    ui.formatReply(ui.showAddTaskMessage(t, tasks.getTasks().size()));
+                    tasks.saveToStorage(storage);
+                } else if (msg.toUpperCase().startsWith(Command.EVENT.name())) {
+                    Task t = parser.parseEvent();
+                    tasks.addTask(t);
+                    ui.formatReply(ui.showAddTaskMessage(t, tasks.getTasks().size()));
+                    tasks.saveToStorage(storage);
+                } else {
+                    ui.displayError("Unable to understand the command. Please enter a valid command.");
+                }
```
Any particular reason you use if-else statements instead of using a switch statement?

limrik commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,187 @@
+     * @param task The task to be marked as not done.
+     * @return A formatted string confirming the task has been marked as not done.
+     */
+    public String showUnmarkTaskMessage(Task task) {
+        return "OK, I've marked this task as not done yet:\n" + "   " + task;
+    }
+
+    /**
+     * Generates a formatted string for marking a task as done and returns the result.
+     *
+     * @param task The task to be marked as done.
+     * @return A formatted string confirming the task has been marked as done.
+     */
+    public String showMarkTaskMessage(Task task) {
+        return "Nice! I've marked this task as done:\n" + "   " + task;
+    }
+
+    /**
+     * Generates a formatted string for adding a task and returns the result.
+     *
+     * @param task The task that has been added.
+     * @param taskCount The current count of tasks in the list.
+     * @return A formatted string confirming the addition of the task.
+     */
+    public String showAddTaskMessage(Task task, int taskCount) {
+        return "Got it. I've added this task:\n" + "   " + task + "\n"
+                + " Now you have " + taskCount + " task(s) in the list.";
+    }
+
+    /**
+     * Generates a formatted string for deleting a task and returns the result.
+     *
+     * @param task The task that has been deleted.
+     * @param taskCount The current count of tasks in the list.
+     * @return A formatted string confirming the deletion of the task.
+     */
+    public String showDeleteTaskMessage(Task task, int taskCount) {
+        return "Noted. I've removed this task:\n" + "   " + task + "\n"
+                + " Now you have " + taskCount + " task(s) in the list.";
+    }
```
Maybe you could consider storing these as format strings instead of methods, can refer to the way AB2 does it :)

limrik commented on src/test/java/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,96 @@
+public class TaskListTest {
+
+    @Test
+    public void addTask_validInput_addsTaskSuccessfully() {
```
Good naming of test methods here!

limrik commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,121 @@
+     * @return True if this object is the same as the obj argument; false otherwise.
+     */
+    @Override
+    public boolean equals(Object obj) {
```
Maybe this could have a more boolean name like isEquals?

limrik reviewed (commented):
Overall nice work! Just some nits on the naming conventions and suggestions.

H4mes commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,161 @@
+    /**
+     * Indicates whether the warning about potential overwriting has been printed.
+     */
+    private boolean warningPrinted = false;
```
Perhaps this could sound more boolean if named "isWarningPrinted"?

H4mes commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,196 @@
+     * @throws DukeException If the command is in an invalid format.
+     */
+    public int parseDelete() throws DukeException {
+        String[] part = command.split(" ");
```
Should this be renamed `parts`? I noticed a change in the variable naming for the next few methods as well.

H4mes commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,196 @@
+                return new Event(description, from, to);
+            }
+        }
+        throw new DukeException("Invalid format. Please use 'event <description> /from <datetime> /to <datetime>'.\n" +
+                " Datetime format: <yyyy-MM-dd HH:mm>.");
```
Perhaps the `+` operator could be moved to the front of the second line to better follow the coding standard? I noticed this in other places too.

H4mes commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,161 @@
+        boolean isDone = Integer.parseInt(parts[1]) == 1;
+        String description = parts[2];
+
+        switch (taskType) {
+            case "T":
+                Task tT = new Todo(description);
+                tT.setDone(isDone);
+                return tT;
+            case "D":
+                Task tD = new Deadline(description, parts[3]);
+                tD.setDone(isDone);
+                return tD;
+            case "E":
+                String[] time = parts[3].split(" - ");
+                if (!(time.length == 2)) {
+                    throw new DukeException("Invalid time format. " +
+                            "Please use <datetime> - <datetime>");
+                }
+
+                Task tE = new Event(description, time[0], time[1]);
+                tE.setDone(isDone);
+                return tE;
+            default:
+                throw new DukeException("Invalid task type " + taskType);
+        }
```
Maybe you should try to follow the switch/case convention in the [layout guide](https://se-education.org/guides/conventions/java/intermediate.html#layout)?

H4mes commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,187 @@
+ */
+public class Ui {
+
+    /**
+     * Scanner object for reading user input.
+     */
+    private final Scanner scanner;
+
+    /**
+     * Horizontal line used for formatting output.
+     */
+    private final String horizontalLine = "____________________________________________________________\n";
```
Should these be in SCREAMING_SNAKE_CASE? Noticed this with several `final` constants.

H4mes commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,121 @@
+     */
+    protected DateTimeFormatter output = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
+
+    /**
+     * Constructs an `Event` with the given description, start time, and end time.
+     *
+     * @param description The description of the event.
+     * @param from The start time of the event in yyyy-MM-dd HH:mm format.
+     * @param to The end time of the event in yyyy-MM-dd HH:mm format.
+     * @throws DukeException If there is an issue parsing the time or the format is invalid.
+     */
+    public Event(String description, String from, String to) throws DukeException {
+        super(description);
+        try {
+            LocalDateTime date1 = LocalDateTime.parse(from, input);
+            LocalDateTime end1 = LocalDateTime.parse(to, input);
+            this.from = date1.format(output);
+            this.to = end1.format(output);
+
+            if (!isBeforeDateTime(date1, end1)) {
+                throw new DukeException("Invalid date range. 'From' should be before 'To'.");
+            }
+        } catch (DateTimeParseException | StringIndexOutOfBoundsException e) {
+            throw new DukeException("Invalid time format. Please use yyyy-MM-dd HH:mm.");
+        }
+    }
```
I like how clean this code is 👍 Also great that you catch errors when the dates are not in the correct order!

H4mes reviewed (commented):
Overall very neat and well-documented code! Just some nits to fix.

felixchanyy committed:
Use Checkstyle for Java code quality checks

The current situation lacks consistent code quality checks, which
can lead to inconsistencies and potential bugs.

Integrate Checkstyle for enforcing coding standards and improving
code quality.

This approach ensures that all code follows the same style guidelines,
enhancing readability and maintainability.

Checkstyle configurations have been added to the project for consistent
code formatting and style checks.

felixchanyy committed:
Merge branch 'branch-A-CheckStyle'

felixchanyy committed:
Learns JavaFX basics

felixchanyy committed:
Create GUI for my chatbot

The current command-line interface for the chatbot is not user-friendly
and lacks visual appeal, making it difficult for users to interact
with the chatbot effectively.

Implement a graphical user interface (GUI) for the chatbot to enhance
user experience and improve usability.

The GUI provides a more intuitive way for users to interact with the
chatbot by presenting the conversation in a visually appealing manner.
It allows users to easily input messages and view responses in a
structured format, enhancing the overall user experience.

This change aligns with the goal of improving user engagement and
satisfaction with the chatbot application.

felixchanyy committed:
Merge branch 'branch-Level-10'

felixchanyy committed:
Generate Duke-GUI.jar file using Gradle

felixchanyy committed:
Refactor files from Duke to Denify

Fixes naming inconsistency in project files

Rename all occurrences of "Duke" to "Denify" across the project files.

This change is necessary to maintain consistency and clarity in the
 project's naming conventions.

- Renamed classes, packages, and resources from "Duke" to "Denify"
for better coherence.
- Updated references and imports to reflect the new naming convention.

Additional context:
- Update imports, package names, and documentation accordingly
- Verify and test the functionality after renaming to ensure no
regressions occur

felixchanyy committed:
Use Assertions

Assert statements are essential for enhancing code reliability and
facilitating debugging processes.

- Add assert statements to critical sections of the codebase to enforce
 expected conditions.
- Ensure that the application behaves as intended by validating inputs,
outputs, and intermediate states.

This approach is chosen because it allows for early detection of
potential issues during development and testing phases. Assertions
serve as a form of self-documentation, providing clarity on expected
behavior and assumptions made within the code.

Additional context:
- Review existing code to identify areas where assert statements
can be beneficially applied.
- Balance assert usage to avoid excessive overhead.

felixchanyy committed:
Merge pull request #1 from felixchanyy/branch-A-Assertions

Use Assertions

felixchanyy committed:
Merge branch 'master' of https://github.com/felixchanyy/ip

felixchanyy committed:
Merge branch 'master' into branch-A-Assertions

felixchanyy committed:
Merge branch 'master' into branch-A-CodeQuality

felixchanyy committed:
Improve code quality

Enhance readability and maintainability by updating variable to standard
naming conventions. This ensures consistency across the codebase
and improves code comprehension for developers.

Rename variables using descriptive nouns and verbs to accurately
represent their purpose and behavior. Follow established naming
conventions for better code organization and understanding.

felixchanyy committed:
Merge pull request #2 from felixchanyy/branch-A-CodeQuality

Branch a code quality

felixchanyy committed:
Merge branch 'master' of https://github.com/felixchanyy/ip

felixchanyy committed:
Merge branch 'master' into branch-A-CodeQuality

felixchanyy committed:
Ensure method/class header comments follow the format
specified in the coding standard

The current situation is that method/class header comments
do not consistently adhere to the specified format, leading to
inconsistency and potential confusion for developers.

To maintain code consistency and improve readability,
this commit ensures that all method/class header comments
follow the prescribed format, particularly focusing on the
 phrasing of the overview statement.

This is done to enforce uniformity across the codebase,
making it easier for developers to understand the purpose
and functionality of methods and classes at a glance.

No other relevant information at this time.

felixchanyy committed:
Merge branch 'branch-A-CodeQuality'

felixchanyy committed:
Merge branch 'branch-A-Assertions'

felixchanyy committed:
Deal with duplicate items -- improve task addition handling

This change is necessary to prevent the addition of duplicate tasks
to the task list. By checking for duplicates before adding a task, I
ensure data consistency and provide better user feedback.

- Modified the addTask method in the TaskList class to check for
duplicate tasks before adding.
- If a duplicate task is found, a DenifyException is thrown with
the message "Task already exists. Delete it to edit."

This approach is chosen to maintain the integrity of the task list
and provide clear guidance to users when attempting to add duplicate
tasks.

felixchanyy committed:
Merge branch 'branch-BCD-Extension'

felixchanyy committed:
Create gradle.yml

Adding Continuous Integration (CI) is necessary to automate 
the build and testing process.

GitHub Actions workflow file gradle.yml is added 
to the repository to set up CI.

This approach is chosen because GitHub Actions provides
a convenient way to define workflows for various CI tasks such as
building, testing, and deployment.

No additional relevant information at this time.

felixchanyy committed:
Update gradle.yml

felixchanyy committed:
Update gradle.yml

felixchanyy committed:
Update gradle.yml

felixchanyy committed:
Improve GUI

Adjusted font sizes, colors and crop display picture for better
readability and added GUI title.

This change is necessary to provide a more cohesive and aesthetically
pleasing user experience. The adjustments ensure that elements are
visually aligned and styled consistently throughout the application,
which enhances usability and reduces cognitive load for users.

By adjusting font sizes and colors, users can more easily navigate
and interact with the graphical user interface. Consistency in design
also helps to reinforce branding and improves overall user satisfaction.

felixchanyy committed:
Add User Guide

To provide users with comprehensive documentation on how to use
the application.

Created a user guide file with detailed instructions and examples
for various features.

This approach ensures that users have access to clear and
concise guidance on using the application effectively.
